sip/sip-002-smart-contract-language.md

```diff
@@ -0,0 +1,309 @@
+  associated with a given key in the map, or returns `'null` if there
+  is no such value.
+2. `(set-entry! map-name key-tuple value-tuple)` - This will set the
+  value of `key-tuple` in the data map
+3. `(insert-entry! map-name key-tuple value-tuple)` - This will set
+  the value of `key-tuple` in the data map if and only if an entry
+  does not already exist.
+4. `(delete-entry! map-name key-tuple)` - This will delete `key-tuple`
+   from the data map
+
+We chose to use data maps as opposed to other data structures for two
+reasons:
+
+1. The simplicity of data maps allows for both a simple
+implementation within the VM, and easier reasoning about
+transactions. By inspecting a given transaction definition, it is
+clear which maps will be modified and even within those maps, which
+keys are affected by a given transaction.
+2. The interface of data maps ensures that the return types of map
+operations are _fixed length_, which is a requirement for static
+analysis of smart contracts' runtime, costs, and other properties.
+
+A smart contract defines the data schema of a data map with the
+`defmap` call. The `defmap` function may only be called in the
+top-level of the smart-contract (similar to `define`). This
+function accepts a name for the map, and a definition of the structure
+of the key and value types. Each of these is a list of `(name, type)`
+pairs, and they specify the input and output type of `fetch-entry`.
+Types are either the values `'principal`, `'integer`, `'bool` or
+the output of a call to `(buffer n)`, which defines an n-byte
+fixed-length buffer. 
+
+### Record Type Syntax
+
+To support the use of _named_ fields in keys and values, our language
+allows the construction of named tuples using a function `(tuple ...)`,
+e.g.,
+
+```
+(define imaginary-number-a (tuple #real 1 #i 2))
+(define imaginary-number-b (tuple #real 2 #i 3))
+
+```
+
+This allows for creating named tuples on the fly, which is useful for
+data maps where the keys and values are themselves named tuples. To
+access a named value of a given tuple, the function `(get #name
+tuple)` will return that item from the tuple.
+\subsection{Principals for Signature Validation}
+
+We provide an additional primitive function \texttt{signed-by?} for
+performing signature validation. This function returns true if the
+given \textit{principal} signed the top-level transaction. A principal
+is a string representation of an entity capable of producing a
+verifiable signature. For example, in Bitcoin, principals are Bitcoin
+addresses, and they could be a RIPEMD-160 hash of a public-key, or a
+script-hash. We use the notion of principals to enable support at the
+protocol level for many different kinds of signatories. For example,
+if we want to continue to support multi-signature principals, the
+protocol could support the standard Bitcoin p2sh multi-sig addresses,
+and support for this would seamlessly be included in any contract.
```

Comment:
I think we need to be careful here in what a `principal` can be, since it seems that `signed-by?` would take a variable amount of time and space depending on what the `principal` is.  For example, if we allow a `principal` to be a whole other contract that gets evaluated, and possibly calls into other contracts in the process, a smart contract developer could accidentally render their script un-runnable because the `signed-by?` operation ends up taking too many resources to run (especially if the execution of `signed-by?` depends on data in a map somewhere that other people store in some other contract).

I'd be in favor of restricting `principal` to be a set of signatures for a contract-defined threshold of m-of-n public keys.  That way, the `signed-by?` method takes a fixed contract-defined amount of resources to evaluate no matter what.

Reply:
I was thinking that the principal would be basically a stacks address, as defined today, and that the signature verification would occur before the transaction ever began processing, such that the smart contract itself is basically just fetching the principal who sent the transaction.

Reply:
So we'd roughly support a pubkey hash, and m-of-n public keys represented as a scripthash.

Reply:
Ah, okay, that makes sense.  `signed-by?` is basically equivalent to comparing a principal identifier to a global, immutable `msg.sender` (in the solidity parlance) and returning true or false.

Reply:
Updated this to add a `tx-sender` field in the latest draft. This should make it more clear that it's pretty much akin to the `msg.sender`